File: scripts/spotify-auth.js
import express from 'express';
import axios from 'axios';
import open from 'open';

if (process.argv.length < 5) {
  console.error(JSON.stringify({
    error: "Missing arguments",
    required: ["clientId", "clientSecret", "redirectUri"]
  }));
  process.exit(1);
}

const [clientId, clientSecret, redirectUri] = process.argv.slice(2);

const app = express();
const PORT = 8888;

app.get('/callback', async (req, res) => {
  try {
    const response = await axios.post('https://accounts.spotify.com/api/token', null, {
      params: {
        grant_type: 'authorization_code',
        code: req.query.code,
        redirect_uri: redirectUri,
        client_id: clientId,
        client_secret: clientSecret,
      },
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded'
      }
    });

    res.send('<script>window.close();</script><p>Autenticação concluída! Esta janela pode ser fechada.</p>');
    console.log(JSON.stringify({
      access_token: response.data.access_token,
      expires_in: response.data.expires_in,
      token_type: response.data.token_type
    }));
    setTimeout(() => process.exit(0), 1000);
  } catch (error) {
    console.error(JSON.stringify({
      error: "Authentication failed",
      details: error.response?.data || error.message
    }));
    process.exit(1);
  }
});

const server = app.listen(PORT, () => {
  const authUrl = `https://accounts.spotify.com/authorize?response_type=code&client_id=${clientId}&scope=user-modify-playback-state&redirect_uri=${encodeURIComponent(redirectUri)}`;
  open(authUrl);
});

server.on('error', (err) => {
  if (err.code === 'EADDRINUSE') {
    console.error(JSON.stringify({
      error: "Port already in use",
      solution: `Close other applications using port ${PORT} or change the port in the script`
    }));
  } else {
    console.error(JSON.stringify({
      error: "Server error",
      details: err.message
    }));
  }
  process.exit(1);
});

File: lua/spotify/spotify-auth.lua
local Utils = require("spotify.utils")
local http = require("plenary.curl")

local redirect_uri = "http://127.0.0.1:8888/callback"
local spotify_auth_url = "https://accounts.spotify.com/api/token"
local auth_success_html = "<script>window.close();</script><p>Autenticação concluída! Esta janela pode ser fechada.</p>"

local function open_browser(client_id, redirect_uri)
    local auth_url = string.format(
        "https://accounts.spotify.com/authorize?response_type=code&client_id=%s&scope=%s&redirect_uri=%s",
        client_id,
        Utils.url_encode("user-modify-playback-state user-read-playback-state"),
        Utils.url_encode(redirect_uri)
    )

    local cmd

    if vim.fn.has("win32") == 1 then
        cmd = { "cmd", "/c", 'start "" "' .. auth_url .. '"' }
    elseif vim.fn.has("mac") == 1 then
        cmd = { "open", auth_url }
    else
        cmd = { "xdg-open", auth_url }
    end

    vim.fn.jobstart(cmd, { detach = true })
end

local function exchange_code(code, client_id, client_secret, redirect_uri, callback)
    local body = {
        grant_type = "authorization_code",
        code = code,
        redirect_uri = redirect_uri,
        client_id = client_id,
        client_secret = client_secret
    }

    local encoded_body = {}
    for k, v in pairs(body) do
        table.insert(encoded_body, string.format("%s=%s", Utils.url_encode(k), Utils.url_encode(v)))
    end

    http.request({
        url = spotify_auth_url,
        method = "POST",
        body = table.concat(encoded_body, "&"),
        headers = {
            ["Content-Type"] = "application/x-www-form-urlencoded",
            ["Accept"] = "application/json"
        },
        callback = function(response)
            if response.status ~= 200 then
                return callback(nil, response)
            end
            local _, data = pcall(vim.json.decode, response.body)
            callback(data)
        end
    })
end

local function start_auth_server(client_id, client_secret, redirect_uri, callback)
    local uri = Utils.parse_uri(redirect_uri)

    local server = vim.loop.new_tcp()
    server:bind(uri.host, uri.port)

    server:listen(128, function(err)
        if err then
            return
        end

        local client = vim.loop.new_tcp()
        server:accept(client)

        local request_data = ""
        client:read_start(function(err, chunk)
            if err then
                client:close()
                return
            end

            if not chunk then
                client:close()
                return
            end

            request_data = request_data .. chunk

            if request_data:find("\r\n\r\n") or request_data:find("\n\n") then
                local pattern = "GET " .. uri.path .. "%?code=([%w-_]+)"
                local code = request_data:match(pattern)

                if code then
                    server:close()

                    local response = string.format(
                        "HTTP/1.1 200 OK\r\n" ..
                        "Content-Type: text/html\r\n" ..
                        "Content-Length: %d\r\n" ..
                        "\r\n%s",
                        #auth_success_html,
                        auth_success_html
                    )

                    client:write(response)

                    vim.defer_fn(function()
                        client:shutdown()
                        client:close()
                        exchange_code(code, client_id, client_secret, redirect_uri, callback)
                    end, 100)
                end
            end
        end)
    end)

    return server
end

local M = {}

---@param credentials Credentials
---@return function
function M.authenticate(credentials, callback)
    local server = start_auth_server(credentials.client_id, credentials.client_secret, redirect_uri, callback)
    open_browser(credentials.client_id, redirect_uri)

    return function()
        if server and not server:is_closing() then
            server:close()
        end
    end
end

return M
